src/createReducer.ts

```diff
@@ -105,12 +105,24 @@ export function createReducer<S>(
       : mapOrBuilderCallback
 
   return function(state = initialState, action): S {
-    // @ts-ignore createNextState() produces an Immutable<Draft<S>> rather
-    // than an Immutable<S>, and TypeScript cannot find out how to reconcile
-    // these two types.
-    return createNextState(state, (draft: Draft<S>) => {
-      const caseReducer = actionsMap[action.type]
-      return caseReducer ? caseReducer(draft, action) : undefined
-    })
+    const caseReducer = actionsMap[action.type]
+    if (caseReducer) {
+      if (isDraft(state)) {
+        // we must already be inside a `createNextState` call, likely because
+        // this is being wrapped in `createReducer` or `createSlice`.
+        // It's safe to just pass the draft to the mutator.
+        const draft = state as Draft<S> // We can aassume this is already a draft
+        return caseReducer(draft, action) || state
```

Comment:
Shouldn't this use nullish coalescing?
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#nullish-coalescing
```suggestion
        return caseReducer(draft, action) ?? state
```

Reply:
Actually come to think of it, this won't work because it's trying to prevent the scenario of the return value being void, but the `?? operator only protects against `null` or `undefined`, but apparently not `void`.

Reply:
Hm. Are you sure about that?
When you have a function that is returning void, you will get `undefined`.

```ts
function foo() {
  return;
}
console.log(foo() ?? "bar");
```
Will log "bar"

Reply:
Here's an example, this should hit some typing errors.
https://www.typescriptlang.org/play/#code/C4TwDgpgBAysCGxoF4oG8BQUoDd4BsBXCALigDtCBbAIwgCcBuDAX2Y3wmCipDkVKwESKAB9cAewCWAEyio0uAsTIBGAAxQ2GDADNC5AMbApE8lCQBnYAAprAsvxHic0mQEpHw6Jmz0uhPTm9iIA-KHoSkSCAExazCw6hmaWEpwAdPgSAOY2AFIwAPIAcunW9FLk2VK6IDZWtrxOEO6tjEA